nvflare/app_common/aggregators/weighted_aggregation_helper.py

```diff
@@ -37,30 +39,32 @@ def reset_stats(self):
 
     def add(self, data, weight, contributor_name, contribution_round):
         """Compute weighted sum and sum of weights."""
-        for k, v in data.items():
-            if self.exclude_vars is not None and self.exclude_vars.search(k):
-                continue
-            weighted_value = v * weight
-            current_total = self.total.get(k, None)
-            if current_total is None:
-                self.total[k] = weighted_value
-                self.counts[k] = weight
-            else:
-                self.total[k] = current_total + weighted_value
-                self.counts[k] = self.counts[k] + weight
-        self.history.append(
-            {
-                "contributor_name": contributor_name,
-                "round": contribution_round,
-                "weight": weight,
-            }
-        )
+        with self.lock:
+            for k, v in data.items():
+                if self.exclude_vars is not None and self.exclude_vars.search(k):
+                    continue
+                weighted_value = v * weight
+                current_total = self.total.get(k, None)
+                if current_total is None:
+                    self.total[k] = weighted_value
+                    self.counts[k] = weight
+                else:
+                    self.total[k] = current_total + weighted_value
+                    self.counts[k] = self.counts[k] + weight
+            self.history.append(
+                {
+                    "contributor_name": contributor_name,
+                    "round": contribution_round,
+                    "weight": weight,
+                }
+            )
 
     def get_result(self):
         """Divide weighted sum by sum of weights."""
-        aggregated_dict = {k: v / self.counts[k] for k, v in self.total.items()}
-        self.reset_stats()
-        return aggregated_dict
+        with self.lock:
+            aggregated_dict = {k: v / self.counts[k] for k, v in self.total.items()}
+            self.reset_stats()
+            return aggregated_dict
```

Comment:
instead of locking it here, 
I think we need to lock it inside "reset_stats" method.
So we don't need lock inside "get_result" method.

Reply:
Incorrect. Line 65 needs to be within the lock since it reads self.counts and self.total and both are modified in "add" method. In fact, this line is where the exception happened.

Reply:
got it, we need to lock L65.

I think we need to lock "reset_stats" method as well.

As it does modify total, counts, history.

Reply:
True, but be careful. If you lock reset_stats() inside of self.lock, then it'll dead lock.

To do it right, you should have a _reset_stats() without lock, and then call it in places with lock.